src/uct/tcp/tcpcm/tcpcm_ep.c

```diff
@@ -0,0 +1,298 @@
+                  "private data. status: %s",
+                  iface, ucs_status_string(hdr->length));
+        return UCS_ERR_IO_ERROR;
+    }
+    hdr->status = UCS_OK;
+    conn_param.private_data_len = sizeof(uct_tcpcm_conn_param_t);
+
+    // send all of the connection info
+
+    //if (UCS_OK != ucs_sys_fcntl_modfl(ep->sock_id_ctx->sock_id, O_NONBLOCK, 0)) {
+    //    ucs_error("fcntl() failed: %m");
+    //    return UCS_ERR_IO_ERROR;
+    //}
+
+    recv_len = recv(ep->sock_id_ctx->sock_id, (char *) &connect_confirm,
+                    sizeof(int), 0);
+    ucs_debug("recv len = %d\n", (int) recv_len);
+    ucs_debug("connect confirm = %d\n", connect_confirm);
+
+    sent_len = send(ep->sock_id_ctx->sock_id, (char *) &conn_param + offset,
+                    (conn_param.private_data_len - offset), 0);
+    ucs_debug("send_len = %d bytes %m", (int) sent_len);
+
+    if (ep != NULL) {
+        pthread_mutex_lock(&ep->ops_mutex);
+        ep->status = UCS_OK;
+        uct_tcpcm_ep_invoke_completions(ep, UCS_OK);
+        pthread_mutex_unlock(&ep->ops_mutex);
+    }
+
+    return UCS_OK;
+}
+
+static inline void uct_tcpcm_ep_add_to_pending(uct_tcpcm_iface_t *iface, uct_tcpcm_ep_t *ep)
+{
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    ucs_list_add_tail(&iface->pending_eps_list, &ep->list_elem);
+    ep->is_on_pending = 1;
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+}
+
+static UCS_CLASS_INIT_FUNC(uct_tcpcm_ep_t, const uct_ep_params_t *params)
+{
+    uct_tcpcm_iface_t *iface       = ucs_derived_of(params->iface,
+                                                    uct_tcpcm_iface_t);
+    const ucs_sock_addr_t *sockaddr = params->sockaddr;
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+
+    UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
+
+    if (iface->is_server) {
+        /* TODO allow an interface to be used both for server and client */
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (!(params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR)) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    UCT_TCPCM_CB_FLAGS_CHECK((params->field_mask &
+                              UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS) ?
+                             params->sockaddr_cb_flags : 0);
+
+    self->pack_cb       = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_SOCKADDR_PACK_CB) ?
+                          params->sockaddr_pack_cb : NULL;
+    self->pack_cb_arg   = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_USER_DATA) ?
+                          params->user_data : NULL;
+    self->pack_cb_flags = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS) ?
+                          params->sockaddr_cb_flags : 0;
+    pthread_mutex_init(&self->ops_mutex, NULL);
+    ucs_queue_head_init(&self->ops);
+
+    /* Save the remote address */
+    if (sockaddr->addr->sa_family == AF_INET) {
```

Comment:
it can be moved to the common code `ucs/sys/sock.[c,h]`

Reply:
you can use `ucs_sockaddr_cmp` from #3435 here

Reply:
Thanks, Dmitry. Will aim to make this change. Should I wait for #3435 to get merged into master or should I go ahead?

Reply:
Akshay, go ahead. No need to wait for this change, this is just for the future.